orttraining/orttraining/test/training_api/common/synthetic_data_loader.cc

```diff
@@ -73,50 +97,41 @@ bool SyntheticDataLoader::GetNextSampleBatch(std::vector<OrtValue*>& batches) {
 
   batches.clear();
 
-  auto memory_info = Ort::MemoryInfo::CreateCpu(OrtArenaAllocator, OrtMemTypeDefault);
+  Ort::MemoryInfo memory_info = Ort::MemoryInfo::CreateCpu(OrtArenaAllocator, OrtMemTypeDefault);
   auto& sample = sample_batch_collections_[sample_batch_iter_index_];
   const auto* ort_api = OrtGetApiBase()->GetApi(ORT_API_VERSION);
-  for (size_t i = 0; i < sample->NumOfInput(); ++i) {
-    auto input_ptr = sample->GetInputAtIndex(i);
-    auto shape_vector = input_ptr->ShapeVector();
-    // Be noted: the created OrtValue won't clean the raw data after its lifetime ended.
-    auto ptr_flt = dynamic_cast<TypedSyntheticInput<float>*>(input_ptr);
-    if (ptr_flt) {
-      OrtValue* value = NULL;
-      ORT_RETURN_ON_ERROR(ort_api->CreateTensorWithDataAsOrtValue(memory_info,
-      input_ptr->GetData<float>().data(), (input_ptr->NumOfElements() * sizeof(float)),
-      shape_vector.data(), shape_vector.size(),
-      ONNXTensorElementDataType::ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT,
-      &value));
-      batches.emplace_back(value);
-      continue;
-    }
-
-    auto ptr_int = dynamic_cast<TypedSyntheticInput<int64_t>*>(input_ptr);
-    if (ptr_int) {
-      OrtValue* value = NULL;
-      ORT_RETURN_ON_ERROR(ort_api->CreateTensorWithDataAsOrtValue(memory_info,
-      input_ptr->GetData<int64_t>().data(), (input_ptr->NumOfElements() * sizeof(int64_t)),
-      shape_vector.data(), shape_vector.size(),
-      ONNXTensorElementDataType::ONNX_TENSOR_ELEMENT_DATA_TYPE_INT64,
-      &value));
-      batches.emplace_back(value);
-      continue;
-    }
-
-    auto ptr_int32 = dynamic_cast<TypedSyntheticInput<int32_t>*>(input_ptr);
-    if (ptr_int32) {
+  for (size_t i = 0; i < sample.NumOfInput(); ++i) {
+    SyntheticInput& input = sample.GetInputAtIndex(i);
+
+    std::visit([&batches, &input, &ort_api, &memory_info](auto&& arg) -> void {
+      ONNXTensorElementDataType elem_data_type;
+      using T = std::decay_t<decltype(arg)>;
+      if constexpr (std::is_same_v<T, std::vector<float>>) {
+        elem_data_type = Ort::TypeToTensorType<float>::type;
+      } else if constexpr (std::is_same_v<T, std::vector<int32_t>>) {
+        elem_data_type = Ort::TypeToTensorType<int32_t>::type;
+      } else if constexpr (std::is_same_v<T, std::vector<int64_t>>) {
+        elem_data_type = Ort::TypeToTensorType<int64_t>::type;
+      } else if constexpr (std::is_same_v<T, std::vector<uint8_t>>) {
+        elem_data_type = ONNXTensorElementDataType::ONNX_TENSOR_ELEMENT_DATA_TYPE_BOOL;
+      } else {
+        throw std::runtime_error("Unsupported element type.");
```

Comment:
nit: it should be limited to the types contained in the variant and I assume those are all supported. if so, it could be something like:
```c++
if constexpr (std::is_same_v<T, std::vector<uint8_t>>) {
  elem_data_type = ONNXTensorElementDataType::ONNX_TENSOR_ELEMENT_DATA_TYPE_BOOL;
} else {
  elem_data_type = Ort::TypeToTensorType<T::value_type>::type;
}
```

Reply:
Should this be a visitor?